### PR TITLE
Allow other datasets in `trainer.evaluate()`

### PR DESCRIPTION
**Pull request by grofte**
In issue https://github.com/huggingface/setfit/issues/365 you mention passing a different dataset to the `evaluate` than the one defined in `eval_dataset=` when you make an instance of `SetFitTrainer`. But you can't do that. Maybe you could before?

So I've added an argument and a test and otherwise only changed the code a little bit. But then I created two commits because I did it all in the browser and I don't know how to rebase (or even edit and save more than one file per commit).

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hey! Thanks for this feature!

> Maybe you could before?

You couldn't - but the Transformers project has a similar interface, and there it is possible.
I made some small changes to prevent some errors:
1. add missing `Dataset` import
2. run `make style` to run automatic formatting

Your implementation & tests look great. I'll merge this once the tests pass.

- Tom Aarsen

**Comment by grofte** [comment]
Bugger, I thought I was clever enough but I still managed to forget the Dataset import. 

Next time I will download the whole thing and do all the things, I promise.

**Comment by grofte** [comment]
I don't think I have much to I want to change now. 

When you add the option for multiple heads, I think a Gaussian mixture model (GMM) head trained with SGD as outlined here would be amazing: https://link.springer.com/article/10.1007/s11063-021-10599-3  
It should basically allow you to estimate whether the data you want to predict on is "in distribution" or "out of distribution". Which again tells us whether we can reasonably expect the model to give a good answer.

If you are completely unfamiliar with GMMs here's a great blog entry: https://koaning.io/posts/feed-forward-posteriors/  
Notice that the neural implementation there doesn't look great (2017 or older), but also keep in mind that expectation maximization for fitting GMMs generally fails with high-dimensionality data.

**Comment by tomaarsen** [comment]
I'm not very familiar with GMM's (although I do know that blogpost author, I spoke with him earlier today haha!), but if there's a `sklearn` implementation, then that could be used instead of the default `LogisticRegression` head, as long as the new head has `fit`, `predict` and `predict_proba` methods.

I just tried using `GaussianMixture()` as the head, and results were certainly not great - but I'm probably using it incorrectly (~1.29% accuracy compared to ~87.17% with log-reg).

- Tom Aarsen

**Comment by grofte** [comment]
Haha, I didn't mean as a classifier head. You might be able to do that but I am fairly certain that is completely impossible with the sklearn implementation when your body has an output with 768 dimensions.

It was more like your own idea where you compare the tSNE for the validation and training data. But by modeling the distribution of activations / body outputs for all training data you should be able to do the same thing for each individual prediction and summarize your tSNE comparison in a single number. 

So one classification head and one training data distribution head.

**Comment by tomaarsen** [comment]
Ohh, I see!

**Comment by grofte** [comment]
That's why I recommended Vincent's blog post. It's only about joint posteriors, not using it as a classifier. SetFit isn't doing anything specific in the contrastive training to give you outputs that look Gaussian. And I'm not sure it should - it should only just change the body enough to allow the log reg to classify successfully, right? Because we are always risking "catastrophic forgetting" when we do these kinds of fine-tunings. And then the ability to generalise outside the few-shot training data gets compromised.

**Comment by tomaarsen** [comment]
> it should only just change the body enough to allow the log reg to classify successfully, right?

That's the goal, yeah - trying to make the embeddings of sentences of the same labels similar, and dissimilar if they have different labels. If successful, logistic regression should be able to solve it without any real issues.
